Task 6

dmitrydi opened the pull request:
pull request for task6 branch

dmitrydi committed:
Create netfix_spec.rb

dmitrydi committed:
Netfix class added

dmitrydi committed:
specs added

dmitrydi committed:
Netfix and MovieToShow added

dmitrydi committed:
netfix_spec and movie_to_show_spec added

dmitrydi committed:
movie_to_show with spec added

dmitrydi committed:
conflic resolved

dmitrydi committed:
beta version of Netfix

dmitrydi committed:
README.md deleted

zverok commented on task_6/spec/netfix_spec.rb:
```diff
@@ -0,0 +1,144 @@
+describe Netfix do
+  before :each do
+    @netfix = Netfix.read
+  end
```
Нужно освоить и использовать [let](http://betterspecs.org/ru/#let) и [subject](http://betterspecs.org/ru/#subject) вместо задания переменных в before-блоке.

zverok commented on task_6/spec/netfix_spec.rb:
```diff
@@ -0,0 +1,144 @@
+
+    it 'creates an array of instances of MovieToShow' do
+      expect(Netfix.read.collection.first).to be_an_instance_of MovieToShow
+    end
```
Эти два теста описывают функциональность, унаследованную от MovieCollection — значит, на примере MovieCollection её и надо тестировать.

zverok commented on task_6/spec/netfix_spec.rb:
```diff
@@ -0,0 +1,144 @@
+    end
+
+    it 'insures @collection items to have :period attribute' do
+      expect(Netfix.read.collection.first.period).not_to be_nil
```
А этот тест описывает функциональность MovieToShow (что у него обязательно должен быть период) — поэтому достаточно тестов того класса.

zverok commented on task_6/spec/netfix_spec.rb:
```diff
@@ -0,0 +1,144 @@
+    it 'should ensure @collection is not empty' do
+      expect(@netfix.filter(period: 'classic').collection).not_to be_empty
+    end
+  end
```
Это тоже уходит в спеку для MovieCollection.
Но вот в Netflix я бы тогда добавил спеку, что оно возвращает an_instance_of Netflix ;)

dmitrydi replied:
да мне особо нигде не надо, чтобы `netfix.filter` был экземпляром `Netfix`. Или это ошибочное мнение?)

zverok replied:
Ну, это было скорее аккуратности ради (родительский класс возвращает из фильтра себя, так что и наследнику хорошо бы тоже себя возвращать), так-то конечно да, можно забить.

zverok commented on task_6/spec/netfix_spec.rb:
```diff
@@ -0,0 +1,144 @@
+      netfix.pay(10)
+      expect(netfix.money).to eq(10)
+      netfix.pay(10)
+      expect(netfix.money).to eq(20)
```
Есть хороший матчер, который так и называется, как описание твоего it: [expect to change](https://www.relishapp.com/rspec/rspec-expectations/v/2-2/docs/matchers/expect-change).

Кроме того, глянь на вот это [правило](http://betterspecs.org/ru/#single) — оно не догма, но позволяет сделать тесты более читаемыми/атомарно тестируемыми.

zverok commented on task_6/spec/netfix_spec.rb:
```diff
@@ -0,0 +1,144 @@
+  describe '#make_shortlist' do
+
+    it 'returns an array of MovieToShow instances' do
+      @netfix.pay(10)
```
А вот это должно бы уйти в before

zverok commented on task_6/spec/netfix_spec.rb:
```diff
@@ -0,0 +1,144 @@
+
+    it 'returns empty array for bad filter' do
+      @netfix.pay(10)
+      expect{@netfix.make_shortlist(genre: 'Tragedy')}.to raise_error(ArgumentError)
```
Поведение правильное, а вот описание привирает :)
В принципе, следуя правилу «одна проверка на тест», можно вообще не писать описаний:

``` ruby
it { expect{@netfix.make_shortlist(genre: 'Tragedy')}.to raise_error(ArgumentError) }
```

zverok commented on task_6/spec/netfix_spec.rb:
```diff
@@ -0,0 +1,144 @@
+      expect{netfix.make_shortlist(period: 'classic')}.to raise_error(RuntimeError, "you don't have enough money")
+    end
+
+    it 'returns an array of MovieToShow instances when paid' do
```
Тест с таким точно описанием уже есть выше.

zverok commented on task_6/spec/netfix_spec.rb:
```diff
@@ -0,0 +1,144 @@
+
+  describe '#show' do
+    it 'may take one argument' do
+      expect(@netfix).to respond_to(:show).with(1).argument
```
На самом деле настолько простые тесты можно и не писать (ты всё равно в тестах посложнее протестируешь, что метод есть, и принимает аргументы).

dmitrydi committed:
classes for diff movie periods added

dmitrydi committed:
specs for new classes added

zverok commented on task_6/classic_movie.rb:
```diff
@@ -0,0 +1,15 @@
+    super(record, host)
+    if !@year.to_i.between?(1946,1968)
+      raise ArgumentError, "year should be in range 1946..1968"
+    end
```
Тебе не кажется, что эти проверки у всех классов подозрительно одинаковые? :)

dmitrydi replied:
...и стоит это запихнуть в какой-нить метод?

zverok replied:
ага. причём метод родителя. а в потомках просто вызывать в конструкторе как-то типа `check_year!(1946, 1968)`

другой подход (выбери тот, который тебе больше нравится) — вообще всю проверку выполнять в родителе, а в потомках просто сделать константу YEAR_RANGE, каждому свою, и родитель пусть её использует при проверке.

dmitrydi replied:
я уже реализовал первый подход

zverok replied:
ага, ок, просто сначала разгрёб вопросы в почте, теперь буду коммиты смотреть :)

zverok commented on task_6/movie_to_show.rb:
```diff
@@ -0,0 +1,70 @@
+                  'new'
+                else
+                  nil
+              end
```
Это ж теперь можно прибрать.

dmitrydi replied:
т.е. вообще не имеет смысла оставлять определение `@period` и `@price` в этом классе?

zverok replied:
Exactly.

zverok commented on task_6/movie_to_show.rb:
```diff
@@ -0,0 +1,70 @@
+                when 2001..Date.today.year
+                  'new'
+                else
+                  nil
```
А вот в таких случаях лучше исключение кидать, а не nil возвращать. информативное исключение точно покажет где баг.

zverok commented on task_6/movie_to_show.rb:
```diff
@@ -0,0 +1,70 @@
+                  5
+                else
+                  nil
+              end 
```
Это должны о себе сами классы-наследники знать.

zverok commented on task_6/movie_to_show.rb:
```diff
@@ -0,0 +1,70 @@
+      when 2001..Date.today.year
+        NewMovie.new(record, host)
+      else
+        MovieToShow.new(record, host)
```
тоже бы исключение.

zverok commented on task_6/netfix.rb:
```diff
@@ -0,0 +1,68 @@
+
+  def read(filename = "movies.txt")
+    @collection = CSV.read(filename, col_sep: "|").map {|a| MovieToShow.create(a, self)}
+    @collection.delete_if {|a| a.class == MovieToShow }
```
это лишнее :) в смысле, так «возможные ошибки» не обходят.

zverok commented on task_6/spec/ancient_movie_spec.rb:
```diff
@@ -0,0 +1,19 @@
+  it{ expect(AncientMovie.new(record)).to be_an_instance_of AncientMovie }
+  it{ expect{AncientMovie.new(bad_record)}.to raise_error(ArgumentError) }
+
+  it{ expect(ancient_movie.to_s).to include("old movie") }
```
:+1:

zverok commented on task_6/spec/classic_movie_spec.rb:
```diff
@@ -0,0 +1,15 @@
+require 'csv'
+
+
+describe ClassicMovie do
```
а вот тут повторения пошли абсолютно таких же тестов, нехорошо. DRY — Don't Repeat Yourself, remember?

dmitrydi replied:
а что можно вместо этого сделать? выписать все в отдельный метод?

zverok replied:
см. тут: https://www.relishapp.com/rspec/rspec-core/docs/example-groups/shared-examples и там: http://betterspecs.org/ru/#sharedexamples

(betterspecs.org вообще рекомендую прочитать от начала до конца, там на час чтения, и сильно помогает понять всё необходимое)

dmitrydi replied:
``` ruby
require_relative '..\movie_classes'
require 'csv'

RSpec.shared_examples "a movie with right :period, :price and #to_s" do |start_year, end_year, expected_period, expected_price, expected_string|
  let(:record) {CSV.read('..\movies.txt', col_sep: "|").map.find{|a| a[2].to_i.between?(start_year, end_year)}}
  let(:bad_record) {CSV.read('..\movies.txt', col_sep: "|").map.find{|a| a[2].to_i > end_year}}
  let(:movie) {described_class.new(record)}

  it{ expect(record[2].to_i).to be_between(start_year, end_year).inclusive }
  it{ expect(bad_record[2].to_i).not_to be_between(start_year, end_year).inclusive }

  it{ expect{described_class.new(bad_record)}.to raise_error(ArgumentError) }

  subject {movie}

  it{ is_expected.to be_an_instance_of described_class }
  it{ is_expected.to have_attributes(:period => expected_period, :price => expected_price)}

  describe '#to_s' do
    subject {movie.to_s}

    it{ is_expected.to include(expected_string) }
  end
end
```

dmitrydi replied:
тест для класса теперь такой

``` ruby
require_relative 'movies_shared_spec'

describe AncientMovie do
  it_behaves_like "a movie with right :period, :price and #to_s", 1900, 1945, 'ancient', 1, 'old movie'
end
```

zverok replied:
Зашибись! Только группы shared examples по смыслу лучше называть, а не описательно. То есть как-то типа `it_behaves_like "movie with limited period"` или чё-то такое.

zverok replied:
И вот, кстати, проверку чего его `to_s` выдаёт, в shared группу лучше не прятать :) Хотя это уже дело вкуса, но тут мы тестируем разное поведение, а не сходное. (Тем более что твой тест пока очень наивен, он не описывает полностью, что мы ожидаем иметь выведенным в `to_s`)

dmitrydi replied:
да, я думал, что проверка to_s относится к другой категории по смыслу. лучше-ка я его выведу отдельно.

dmitrydi replied:
а вот так:

``` ruby
require_relative '..\movie_classes'
require 'csv'

RSpec.shared_examples "a movie with limited period and certain price" do |start_year, end_year, expected_period, expected_price|
  let(:record) {CSV.read('..\movies.txt', col_sep: "|").map.find{|a| a[2].to_i.between?(start_year, end_year)}}
  let(:bad_record) {CSV.read('..\movies.txt', col_sep: "|").map.find{|a| (a[2].to_i > end_year || a[2].to_i < start_year)}}
  let(:movie) {described_class.new(record)}

  it{ expect(record[2].to_i).to be_between(start_year, end_year).inclusive }
  it{ expect(bad_record[2].to_i).not_to be_between(start_year, end_year).inclusive }

  it{ expect{described_class.new(bad_record)}.to raise_error(ArgumentError) }

  subject {movie}

  it{ is_expected.to be_an_instance_of described_class }
  it{ is_expected.to have_attributes(:period => expected_period, :price => expected_price)}
end

RSpec.shared_examples "prepare an instance" do |start_year, end_year|
  let(:record) {CSV.read('..\movies.txt', col_sep: "|").map.find{|a| a[2].to_i.between?(start_year, end_year)}}
  let(:movie) {described_class.new(record)}
end
```

``` ruby
require_relative 'movies_shared_spec'

describe AncientMovie do

  it_behaves_like "a movie with limited period and certain price", 1900, 1945, 'ancient', 1

  describe '#to_s' do
    include_examples "prepare an instance", 1900, 1945
    it { expect(movie.to_s).to eq("#{movie.title} - old movie (#{movie.year})") }
  end

end
```

zverok replied:
Так норм.! (Хотя отдельный shared prepare an instance мне не очень нравится, но мы об этом ещё поговорим, когда общая картина будет виднее.)

(но такие длинные примеры лучше просто коммитить сразу, и в коммитах обсуждать, там бы я смог построчно прокомментировать, если бы было надо)

dmitrydi replied:
ок, принято. я тогда дальше пойду пока.

zverok commented on task_6/theatre.rb:
```diff
@@ -0,0 +1,21 @@
+require_relative 'netfix'
+require 'time'
+
+class Theatre < Netfix
```
Не, кинотеатр не может быть наследником Нетфликса.

zverok commented on task_6/spec/theatre_spec.rb:
```diff
@@ -0,0 +1,24 @@
+describe Theatre do
+  before :each do
+    @theatre = Theatre.read
+  end
```
subject

dmitrydi replied:
это из старого осталось, я класс Theatre пока похерю.

zverok commented on task_6/spec/theatre_spec.rb:
```diff
@@ -0,0 +1,24 @@
+    expect(@theatre).to respond_to(:collection)
+    expect(@theatre.collection).to be_an_instance_of Array
+    expect(@theatre.collection).not_to be_empty
+    expect(@theatre.collection.first).to be_an_instance_of MovieToShow
```
Если используешь subject, сможешь писать вот так:

``` ruby
it { is_expected.to have_attributes.... }
```

а добавив гем rspec-its — и вот так:

``` ruby
its(:collection) { is_expected.to be_an Array }
```

dmitrydi replied:
кстати, про subject - он определяется раз и навсегда для всего теста? когда я пишу

``` ruby
subject {ancient_movie}

  it{ is_expected.to be_an_instance_of AncientMovie }
  it{ is_expected.to have_attributes(:period => 'ancient', :price => 1)}

  subject {ancient_movie.to_s}

  it{ is_expected.to include("old movie") }
```

то выдает ошибку, связанную с тем, что subject становится String'ом, соотв, две первые проверки не проходят

zverok replied:
subject определяется для текущего контекста. То есть вот что ты хочешь, на самом деле:

``` ruby
context 'AncientMovie' do
  subject {ancient_movie}

  it{ is_expected.to be_an_instance_of AncientMovie }
  it{ is_expected.to have_attributes(:period => 'ancient', :price => 1)}

  context '#to_s' do
    subject {ancient_movie.to_s}

    it{ is_expected.to include("old movie") }
  end
end
```

вместо слова context можно использовать describe, оно даёт немножко полезных плюшек, вот так:

``` ruby
describe AncientMovie do
  subject {ancient_movie}

  it{ is_expected.to be_an_instance_of described_class }
  it{ is_expected.to have_attributes(:period => 'ancient', :price => 1)}

  context '#to_s' do
    subject {ancient_movie.to_s}

    it{ is_expected.to include("old movie") }
  end
end
```

dmitrydi replied:
такое исполнение проверок для класса норм?

``` ruby
require_relative '..\ancient_movie'
require 'csv'

describe AncientMovie do
  let(:record) {CSV.read("movies.txt", col_sep: "|").map.find{|a| a[2].to_i.between?(1900,1945)}}
  let(:bad_record) {CSV.read("movies.txt", col_sep: "|").map.find{|a| a[2].to_i.between?(1946,1968)}}
  let(:ancient_movie) {AncientMovie.new(record)}

  it{ expect(record[2].to_i).to be_between(1900,1945).inclusive }
  it{ expect(bad_record[2].to_i).not_to be_between(1900,1945).inclusive }

  it{ expect{AncientMovie.new(bad_record)}.to raise_error(ArgumentError) }

  subject {ancient_movie}

  it{ is_expected.to be_an_instance_of AncientMovie }
  it{ is_expected.to have_attributes(:period => 'ancient', :price => 1)}

  describe '#to_s' do
    subject {ancient_movie.to_s}

    it{ is_expected.to include("old movie") }
  end

end
```

zverok replied:
:+1:

dmitrydi committed:
changes to movie classes

MovieToShow: right_year? added, self.create added, @price and @period vars excluded
other Movie classes: checks added, @price and @period vars initializing added

dmitrydi committed:
some new specs added

shared examples added for movie classes and netfix class, specs for movie classes added

zverok commented on task_6/ancient_movie.rb:
```diff
@@ -0,0 +1,16 @@
+    super(record, host)
+    self.right_year?(1900, 1945)
+    @period = 'ancient'
+    @price = 1
```
Эти ж обе штуки — константы, в чём смысл их в переменных объекта хранить? (И в чём кстати вообще смысл отдельно иметь строку "ancient", как ты её используешь?)

dmitrydi replied:
пока никак не использую, но, например, могу использовать в фильтре класса netfix.filter(period: 'ancient')

zverok replied:
YAGNI :) В смысле, есть хороший принцип you aint gonna need it = «(а вдруг) оно тебе не понадобится» — то есть не добавлять фичек, пока они не нужны (а как понадобятся, можно будет это значение вычислять просто из имени класса, на лету ;))

dmitrydi replied:
типа `AncientMovie.class.to_s.delete("movie")` ?

dmitrydi replied:
я вот немного не пойму. у меня сейчас есть метод `Netfix#filter`. В текущей реализации процедура выбора по жанру и (к примеру) цене тривиальна - фильтр работает с атрибутами `@oeriod и @price`. Насчет цены - понятно, ее разумно делать константой, дабы не поменять случайно. Но если я уберу эти атрибуты, то, как минимум, фильтр перестанет работать для этих атрибутов, как максимум - перестанут работать все остальные методы, где промелькнули фразы типа `.price`. Может тогда имеет смысл объявить в классе `MovieToShow` методы `#price и #period`, которые возвращают соответствующие значения? Интерфейс для остальных методов не изменится.

zverok commented on task_6/movie_to_show.rb:
```diff
@@ -0,0 +1,30 @@
+    super(record, host)
+  end
+
+  def right_year?(min_year, max_year)
```
а) метод, который называется на вопр.знак — должен возвращать true/false и только, а если метод что-то проверяет и кидает ошибку — его называют как-то типа check_year или validate_year (ещё можно добавить воскл.знак, типа `check_year!` — мне нравится, но не все такой стиль считают приличным)
б) идиоматичнее использовать не две отдельных переменных, а range

zverok commented on task_6/movie_to_show.rb:
```diff
@@ -0,0 +1,30 @@
+
+  def self.create(record, host = nil)
+    case record[2].to_i
+      when 1900..1945
```
Обрати внимание, что теперь у тебя эти годы в двух местах написаны — тут и в самом классе во время проверки; хорошо бы их в константу вынести в каждом классе, м? И здесь использовать.

А иначе поменяются требования (например, решим старым считать кино до 40-х годов) — и придётся менять в двух местах, и про одно обязательно забудем.

zverok commented on task_6/netfix.rb:
```diff
@@ -0,0 +1,73 @@
+  end
+
+  def self.read(filename = "movies.txt")
+    Netfix.new.read(filename)
```
Зачем? Уже ж есть этот код в MovieCollection

zverok commented on task_6/netfix.rb:
```diff
@@ -0,0 +1,73 @@
+
+  def films_by_producers
+    @movs_by_producers ? @movs_by_producers : @movs_by_producers = @collection.group_by {|a| a.producer}.map{|key, val| [key,val.map {|a| a.title}]}.to_h
+  end
```
А это нам тут, кстати, зачем?..

dmitrydi replied:
А это для ClassicMovie, чтобы список фильмов режиссера выводить

zverok replied:
Так может оно в MovieCollection пусть будет?..

dmitrydi replied:
я думал, в предыдущие файлы не лазим

zverok replied:
отчего же?.. программный проект цементом не заливают, идея в том что когда у нас есть полезный метод, который должен быть в осн.классе — или не ломающее этот клас изменение функциональности — то пусть там и будет.

zverok commented on task_6/netfix.rb:
```diff
@@ -0,0 +1,73 @@
+  def how_much?(name)
+    begin
+      @collection.find{|a| a.title == name}.price
+    rescue
```
rescue без указания чего именно ожидаешь — дурной тон (в смысле, тут же у тебя один вариант, в котором этот rescue разумен; а при этом перехватит он совершенно неожиданные ошибки, если вдруг у тебя где баг).

но в целом это вообще не оч.красиво — по прилетанию ошибки решать, что фильма не было.

zverok commented on task_6/spec/movie_to_show_spec.rb:
```diff
@@ -0,0 +1,63 @@
+    let(:movie_to_show) {MovieToShow.new(record)}
+    let(:year) {movie_to_show.year.to_i}
+
+    it "returns nil when @year in specified range" do
```
учитывая, что этот метод нужен только для бросания ошибки, лучше проверять `expect().to_not raise_error`

dmitrydi committed:
classes AncientMovie and MovieToShow changed

dmitrydi committed:
Specs for MovieToShow and AncientMovie changed

zverok commented on task_6/ancient_movie.rb:
```diff
@@ -0,0 +1,23 @@
+  PRICE = 1
+
+  def initialize(record, host = nil)
+    super(record, host)
```
Кстати, если super принимает те же аргументы, то можно писать просто `super` (без аргументов, это значит «передать все»).

zverok commented on task_6/ancient_movie.rb:
```diff
@@ -0,0 +1,23 @@
+
+  def initialize(record, host = nil)
+    super(record, host)
+    self.check_year(@year)
```
Впрочем, обрати внимание, что и эту проверку уже можно перенести в конструктор MovieToShow, и отдельного конструктора тут вообще не нужно будет уже.

zverok commented on task_6/movie_to_show.rb:
```diff
@@ -0,0 +1,32 @@
+      when ModernMovie::PERIOD
+        ModernMovie.new(record, host)
+      when NewMovie::PERIOD
+        NewMovie.new(record, host)
```
Теперь этот кейс можно просто в цикл превратить :) `[классы фильмов].detect{период подходит}` и т.д.

dmitrydi replied:
я вот никак не соображу теперь - как можно кинуть исключение, если в записи кривой год, и при этом чтоб код был изящным? сейчас пока вот так, без проверки:

``` ruby
def self.create(record, host = nil)
    [AncientMovie, ClassicMovie, ModernMovie, NewMovie].detect {|clas| clas::PERIOD.include?(record[2].to_i)}.new(record, host)
```

dmitrydi replied:
исключение-то прилетит в случае чего, но это будет NoMethodError же ж.

zverok replied:
``` ruby
clas = [AncientMovie, ClassicMovie, ModernMovie, NewMovie].detect {|clas| clas::PERIOD.include?(record[2].to_i)} or
  raise(.....)
# или отдельной строкой:
raise(...) unless clas
clas.new(data)
```

Ничо изящнее тут не придумаешь. в общем.
(кстати, в Рубишном коде принято писать `klass`, когда надо назвать переменную как-то так)

dmitrydi commented on task_6/spec/ancient_movie_spec.rb:
```diff
@@ -1,12 +1,33 @@
+  	  expect(ancient_movie.price).to eq(1)
+  	end
+  end
 
```
а что насчет методов `#price и #period` - это хороший подход?

zverok replied:
да, вполне. можно сказать «пусть используют константу», но так лучше (метод, оборачивающий константу) — и позволяет в дальнейшем как-то код развивать.

dmitrydi committed:
movie classes chenged

#price and #period methods moved to MovieTpShow

dmitrydi committed:
specs for AncientMovie and MovieToShow changed

zverok commented on task_6/movie_to_show.rb:
```diff
@@ -2,31 +2,35 @@
 
 class MovieToShow < Movie
+  PERIOD = nil
+  PRICE = nil
```
Лучше вообще убрать константы, мне кажется (и сделать таким образом этот класс «абстрактным» — если кто-то попытается создать его экземпляр, он будет просто падать, работать будут только наследники).

dmitrydi replied:
ну да, а так получается как будто недокласс - две левые константы, а по факту используем все равно только наследников.

zverok replied:
Exactly. И код упростится, если не пытаться заставить его работать в таком вот «недо-состоянии».

dmitrydi replied:
и нужно ли тогда в initialize добавить выброс исключения если self.class == MovieToShow?

zverok replied:
не, это перебор. само потом упадёт, если неверно используют

zverok commented on task_6/spec/movie_to_show_spec.rb:
```diff
@@ -11,13 +11,27 @@
+
+    it 'raise an error when year value is not recognized' do
+      bad_record = record
+      bad_record[2] = 1600
```
Лучше через let «плохую» запись создать.

dmitrydi committed:
final (hope) versions of movie classes

dmitrydi committed:
specs for updated movie classes

dmitrydi committed:
ancient, modern, new movie specs unified

dmitrydi committed:
specs update

dmitrydi committed:
netfix and moviecollection update

dmitrydi committed:
specs netfix and moviecollection update

dmitrydi committed:
changes

dmitrydi committed:
last result before holidays

dmitrydi committed:
MovieCollection#filter changed

zverok commented on task_6/movie_collection.rb:
```diff
@@ -56,12 +56,20 @@ def sort_by(key)
+          else
+            a.send(key).include?(val)
+          end
+        }
```
Это начинает быть странным, тебе не кажется?.. Тут бы помог отедльный метод «сравнить значние с фильтром»

zverok replied:
И, кроме того: 
1. `value.class == Fixnum` → `value.is_a?(Fixnum)`
2. `value.is_a?(Fixnum) || value.is_a?(Float)` → `value.is_a?(Numeric)` (у нас тут ООП! поэтому классы с близким смыслом и поведением скорее всего имеют общего родителя)

dmitrydi replied:
``` ruby
# Movie#key_eq_to_val?
  def key_eq_to_val?(key, val)
    if self.is_a?(Numeric)
      self.send(key) == val
    else
      self.send(key).include?(val)
    end
  end
```

dmitrydi replied:
ой, нет

``` ruby
  def key_eq_to_val?(key, val)
    key_val = self.send(key)
    if key_val.is_a?(Numeric)
      key_val == val
    else
      key_val.include?(val)
    end
  end
```

zverok replied:
Ага, вроде того. Ну только название можно и короче. Типа `matches?(key, check)`

zverok commented on task_6/netfix.rb:
```diff
@@ -28,12 +28,22 @@ def make_shortlist(filt = nil)
+    sum_rating_ary = shortlist.map{|a| s += (k*a.rating + b)}
     rnd = Random.new.rand(0..sum_rating_ary.last)
     sum_rating_ary.index {|a| a >= rnd}
   end
```
Сложноват метод-то. Надо проще!

dmitrydi replied:
я сделал так
`movie = list_to_show.max_by{ |a| rand * a.rating }`

zverok replied:
:+1:

dmitrydi committed:
new specs added

dmitrydi committed:
Cinema, Theatre classes added

dmitrydi committed:
theatre spec fixed

dmitrydi committed:
theatre fixed

zverok commented on task_6/spec/cinema_spec.rb:
```diff
@@ -0,0 +1,40 @@
+  end
+
+  describe '#self.read' do
+    it {expect(cinema).to be_instance_of Cinema}
```
`it { is_expected.to be....}` можно будет написать, если сделать его subject-ом.

zverok commented on task_6/spec/cinema_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+  describe '#pay' do
+    it { expect{cinema.pay(-1)}.to raise_error(ArgumentError, "argument should be >=0") }
+    it { expect{cinema.pay(10)}.to change{cinema.money}.from(0).to(10) }
```
:+1:

zverok commented on task_6/spec/cinema_spec.rb:
```diff
@@ -0,0 +1,40 @@
+    it { expect(cinema.pay(10)).to be_instance_of Cinema }
+  end
+
+  describe '#how_much?' do
```
Соблазнительно назвать его именно так, но метод с вопр.знаком по общему соглашению/ожиданию всегда возвращает только true/false. А этот уж лучше звать `price_for(...)`, напр.

dmitrydi replied:
ну, насчет имени метода не я придумал)
"можно спросить, сколько стоит нужное кино: netflix.how_much?('The Terminator');"

zverok replied:
О боже. Я понял, в тот момент когда я это писал — видимо хотел развлечь учеников идеей, что в названии метода может быть вопросительный знак. Сорри, поправлю. Так, конечно, забавно, но нехорошо.

zverok commented on task_6/spec/cinema_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+  describe '#how_much?' do
+    it 'expect to return right price' do
+      price = cinema.filter(title: "The Terminator").collection.first.price
```
Не ленись сделать лишний let, с тебя не убудет :)

zverok commented on task_6/spec/cinema_spec.rb:
```diff
@@ -0,0 +1,40 @@
+  end
+
+  describe '#inspect' do
+    it {expect(cinema.inspect).to be_instance_of Hash }
```
????? `inspect` в Руби всегда должен возвращать строку, иначе он бесполезен.

dmitrydi replied:
у меня сейчас inspect выглядит так

``` ruby
  def inspect
    self.stats(:period)
  end
```

в принципе, работает хорошо при тестировании, но возвращает не строку.

zverok replied:
Смотри, `inspect` — по определению метод, позволяющий получить «какое-то представление об объекте» в строковом представлении. Это позволяет использовать его при работе в irb, бросании ошибок (типа `ArgumentError, "Expected String, received #{obj.inspect}"`) и прочих случаях именно для _вывода на печать_. `inspect`, не возвращающий строку, может сломать что угодно непредсказуемым образом.

zverok commented on task_6/spec/movie_collection_spec.rb:
```diff
@@ -14,12 +20,14 @@
+    it {expect(movie_collection.filter(rating: 8.3).collection).to all(have_attributes(:rating => 8.3)) }
+
+    let(:filtered_collection) {movie_collection.filter(actors: 'Sharon Stone').collection}
+    it {expect(filtered_collection).to all(contain_value('actors', 'Sharon Stone'))}
```
а) раз уж это название метода, лучше бы символ: `contain_value(:actors, 'Sharon Stone')`
б) и кстати, я вспомнил что rspec matchers are cool: `to all have_attributes(actors: array_including('Sharon Stone'))` ;)

dmitrydi replied:
вот где про эти все фичи прочитать? про сам руби понятно где читать, а про rspec немного куцее описание, как будто бы гуглить надо на произвольную тему.

zverok replied:
На самом деле всё (относительно ;)) просто — rspec состоит из довольно немногих концепций:
- собственно сам rspec, его ядро (организация тестов, describe/context/let/subject/it и всё)
- matchers/expectations — собственно, вот эти самые штуки, которые позволяют проверять что угодно коротко и читабельно
- method stubs & mocks — это пока тебе не пригодилось
- (и ещё кое-что, в следующих уроках наткнёмся)

В данном случае, хороший запрос к гуглу — rspec matchers array или что-то такое.
Надо признать, что я немного «схитрил» (но ты успеешь получить подобный опыт, так что всё ок) — сказал же, что не сразу вспомнил, что так можно: https://relishapp.com/rspec/rspec-mocks/v/3-2/docs/setting-constraints/matching-arguments — обычно, такие составные матчеры используются как раз при проверке method stubs, до этого мы дойдём. Но пригождаться могут везде.

В принципе, тут конечно твой custom matcher вполне неплох был, можно игнорировать мой совет :) Только имя его наверное чуть лучше бы как-то вот так: `include_in_attribute(:actors, 'Sharon Stone')` (тогда он будет вполне выглядеть как смесь матчеров include и have_attribute)

dmitrydi committed:
#inspect fixed, #how_much? changed to #price_for

dmitrydi committed:
theatre_spec update

dmitrydi committed:
theatre class update

zverok commented on task_6/book.rb:
```diff
@@ -0,0 +1,9 @@
+            @author = author
+            @category = category
+        end
+end
```
А это нам, к слову, зачем?..

dmitrydi replied:
Сорри, это мусор

zverok replied:
Ну дык повыкидывай мусор-то. А то коммит в целом рассматривать неудобно.

zverok replied:
Удали, говорю, ненужные файлы.

zverok commented on task_6/cinema.rb:
```diff
@@ -0,0 +1,53 @@
+      raise ArgumentError, "no such movie"
+    else
+      movie.price
+    end
```
unless + else — очень дурной тон (потому что в голове получается «если не... иначе» — эээ «иначе в смысле в каком случае???»). И вообще хорошо бы вспомнить, что мы уже разбирали — если исключение в начале выкинули, то дальше никаких else не нужно.

Вообще в таком случае можно использовать метафору с and/or (хотя не все её любят):

``` ruby
movie = @collection.find{|a| a.title == name} or raise ArgumentError, "no such movie"
```

И, наконец, хорошо бы исключения информативнее делать — то есть что-то типа «фильм (имя фильма) не найден»

zverok commented on task_6/cinema.rb:
```diff
@@ -0,0 +1,53 @@
+  attr_reader :money
+
+  def read(filename = "movies.txt")
+    @collection = CSV.read(filename, col_sep: "|").map {|a| MovieToShow.create(a, self)}
```
Это ж повторяет код из MovieCollection#read, не?

dmitrydi replied:
не. там `...map.{ |a| Movie.new(...)}`

zverok replied:
Ага. Но остальная-то вся конструкция повторяется. Нехорошо, не люблю когда код повторяется ;)

dmitrydi replied:
не пойму, как существенно упростить эту историю. `CSV.read(filename, col_sep: "|")` разве что в отдельный метод вынести? там же похожего больше нет ничего.

zverok replied:
dependency injection :) То есть в родительском классе пишешь:

``` ruby
CSV.read(filename, col_sep: "|").map {|a| create_movie(a)}
```

...и тогда во всех потомках можешь переопределять ТОЛЬКО create_movie и всо.

zverok commented on task_6/cinema.rb:
```diff
@@ -0,0 +1,53 @@
+    self
+  end
+
+  def pay(amount)
```
Почему метод в этом классе?

dmitrydi replied:
потому что используется в потомках Cinema - Netfix & Theatre

zverok commented on task_6/cinema.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def inspect
+    self.stats(:period).collect{ |k, v| "#{k} - #{v}"}.join(", ")
```
Лишний `self.`

zverok commented on task_6/cinema.rb:
```diff
@@ -0,0 +1,53 @@
+    self.stats(:period).collect{ |k, v| "#{k} - #{v}"}.join(", ")
+  end
+
+  def filter_by_price(money)
```
А вроде ж MovieCollection умел фильтровать по любому полю?..

dmitrydi replied:
нет, тут смысл в том, что сравнение не `==` или `include?`, а `<=`

zverok replied:
ну дык можно же переформулировать так, чтобы повторно использовать имеющийся метод, а не плодить новые, похожие, «почти такие же». например, `x <= 10` это то же самое, что `(0..10).include?(x)` ;)

dmitrydi replied:
да, но MovieCollection#filter кидает исключение "no movies found with the filter given", а filter_by_price - "you don't have enough money"

zverok replied:
We both would be happy with something like this:

``` ruby
def filter_by_price(value)
  filter(price: 0..value)
rescue MovieNotFound
  raise "You don't have enough money"
end
```

Например. Или иной редизайн (я вижу несколько разных вариантов), уменьшающий дублирование похожего кода.

dmitrydi replied:
я немного не пойму - ты когда пишешь `rescue MovieNotFound`  - это что? в смысле, есть классы ошибок ArgumentError, RuntimeError, и т.д., а `MovieNotFound` - как слепить?

zverok replied:
You can define your own exception class, easily :)
Полезно как раз для таких случаев (когда тебе надо перехватывать именно какую-то определённую ошибку, а не все возможные).

dmitrydi replied:
я думаю вообще отказаться от этого метода и слепить нечто такое

``` ruby
  def select_movie(filter = nil)
    filtered_collection = self.filter(filter).collection
    filtered_collection.select! { |m| m.price <= @money }
    raise "You don't have enough money" if filtered_collection.empty?
    filtered_collection.max_by{ |a| rand * a.rating }
  end
```

zverok replied:
Тоже имеет смысл.

zverok commented on task_6/cinema.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+  end
+
+  def show(filter = nil)
```
Почему метод тут?

dmitrydi replied:
потому что одинаков для потомков

zverok replied:
хм. ну, некий смысл, в этом, в общем есть. но почему лучше его отдельно у потомков оставить — для читабельности кода. потому что отдельно у потомков можно будет назвать аргумент по-разному, и потому что скажем для Theater он не может быть nil. При этом реальную работу может выполнять родительский класс, т.е. в потомках будет типа

``` ruby
def show(time)
 show_movie(select_by_time(time))
end
```

zverok commented on task_6/movie_collection.rb:
```diff
@@ -15,7 +15,7 @@ def read(filename = "movies.txt")
 
   def self.read(filename = "movies.txt")
-    MovieCollection.new.read(filename)
+    self.new.read(filename)
```
лишний self

zverok commented on task_6/movie_collection.rb:
```diff
@@ -31,6 +31,10 @@ def to_s
   end
 
+  def films_by_producers
+    @movs_by_producers ? @movs_by_producers : @movs_by_producers = @collection.group_by {|a| a.producer}.map{|key, val| [key,val.map {|a| a.title}]}.to_h
```
`||=`

zverok replied:
http://www.rubyfleebie.com/check-for-nil-and-initialize-on-a-single-line/

zverok commented on task_6/movie_collection.rb:
```diff
@@ -49,15 +53,25 @@ def sort_by(key)
+      else
+        self.class.new(filtered_collection)
+      end
+    end
```
Это очень запутанно (не говоря уж про unless/else).

dmitrydi replied:
```
  def filter(filt = nil)
    filtered_collection = filt ? @collection.find_all { |a| filt.inject(true) { |memo, (key, val)| memo && a.match?(key, val) } } : @collection
    raise ArgumentError, "No movies found with filter #{filt}" if filtered_collection.length == 0
    self.class.new(filtered_collection)
  end
```

dmitrydi replied:
так норм?

zverok replied:
Неа :) Вопрос же не в количестве строчек, а в логике. Смотри:
1. Если фильтров нету — то можно сразу вернуть значение (т.е. первой строкой `return self unless filters`)
2. Или сделать дефолтным значением аргумента — пустой хеш. Это нам сразу уберёт `?:` в первой строке. По-моему, это даже аккуратнее.
3. inject можно вывернуть наизнанку, будет лучше:

``` ruby
filters.inject(@collection) { |memo, (key, val)| collection.select{|m| m.match?(key, val)} }
```

↑ тут, конешно, всё равно два блока, но внутренний совсем тривиальный, и следить за мыслью проще.

Ну и по мелочи: `filtered_collection.length == 0` → `filtered_collection.empty?`

dmitrydi replied:
слушай, так не работает, либо я что-то не понимаю.

``` ruby
filtered_collection = filt.inject(@collection) { |memo, (k,v)| @collection.select{ |m| m.match?(k,v) } }
```

dmitrydi replied:
в этой версии фильтра же вот что происходит (как я понимаю): для каждого элемента filt из исходной коллекции @collection выбираются те фильмы, для которых .match?(k, v). Сумма этих фильмов возвращается в конце. НО! нам же нужно логическое И между фильтрами, а не ИЛИ, а тут ИЛИ же выходит? Нет?

dmitrydi replied:
а, стоп, вот так должно сработать

``` ruby
filtered_collection = filt.inject(@collection) { |memo, (k,v)| memo.select{ |m| m.match?(k,v) } }
```

dmitrydi replied:
да, сработало

zverok commented on task_6/movie_to_show.rb:
```diff
@@ -0,0 +1,38 @@
+      price_range.include?(self.price)
+    else
+      self.price == price_range
+    end
```
Ээээ ничо не понял. Можешь объяснить логику?

dmitrydi replied:
логика в том, что иногда удобно узнать, в каком диапазоне цена фильма, а иногда - какое конкретно значение. в метод можно передать как диапазон, так и единичное значение

zverok replied:
Хм. Так понятнее.
Но я на него ни спек не вижу, ни использования в другом коде.
И на самом деле ты хочешь тут не if, а оператор `===`:

``` ruby
(10...20) === 15 #=> true
(10...20) === 21 #=> false
15 === 15 # => true
15 === 21 # => false
```

Отличная штука, очень рекомендую :)

zverok replied:
http://zverok.github.io/blog/2015-07-18-matchish.html — почитай в этом посте в моём бложике раздел "Three Equality Signs" :) (да, это скрытая реклама!)

zverok commented on task_6/movie_to_show.rb:
```diff
@@ -0,0 +1,38 @@
+  end
+
+
+  def check_year(year)
```
Я так мыслю, что это приватный метод должен быть.

zverok commented on task_6/spec/cinema_spec.rb:
```diff
@@ -0,0 +1,44 @@
+  end
+
+  describe '#filter_by_price' do
+    it { expect(cinema.pay(3).filter_by_price(cinema.money).collection).to all have_price(1..3)}
```
Вот что-то не так с этим методом. Если показываемое использование «некто должен взять деньги у самого же кинотеатра, и тот же кинотеатр попросить отфильтровать» — то по-моему у нас что-то с разделением ответственности не то.

zverok commented on task_6/spec/movie_to_show_spec_o.rb:
```diff
@@ -0,0 +1,30 @@
+require_relative '../movie_classes'
```
Почему два файла?..

dmitrydi replied:
это снова что-то лишнее попало в кучу

zverok commented on task_6/spec/movies_spec.rb:
```diff
@@ -0,0 +1,17 @@
+require 'csv'
+require 'rspec/expectations'
+
+RSpec::Matchers.define :raise_error_whith_bad_years do |min_year, max_year|
```
Лучше `:check_year_range`. Будет читаться в лоб: `expect(AncientMovie).to check_year_range(1900..1936)`, напр. М?

dmitrydi replied:
это старое осталось. сейчас этот спек вообще не используется. я проверку запихнул в shared examples

zverok commented on task_6/spec/theatre_spec.rb:
```diff
@@ -0,0 +1,38 @@
+  describe '#select_movie' do
+    it{ expect{Theatre.new.select_movie}.to raise_error(ArgumentError) }
+    it{ expect(theatre.select_movie('08:30')).to have_attributes(:period => 'ancient') }
+    it{ expect(theatre.select_movie('12:30').genre).to include('Comedy').or include('Adventure') }
```
а ты ж сделал себе отличный матчер — используй его: `.to include_in_attribute(genre: 'Comedy')`. м?

zverok commented on task_6/spec/theatre_spec.rb:
```diff
@@ -0,0 +1,38 @@
+
+  describe '#show' do
+    it{ expect{theatre.show}.to raise_error(ArgumentError) }
+    it{ expect{Theatre.read.show('08:30')}.to raise_error(RuntimeError, "you don't have enough money")}
```
Вроде мы не договаривались про деньги в обычном кинотеатре :)

dmitrydi replied:
хм. йа этого не просёк.

zverok commented on task_6/spec/theatre_spec.rb:
```diff
@@ -0,0 +1,38 @@
+  describe '#show' do
+    it{ expect{theatre.show}.to raise_error(ArgumentError) }
+    it{ expect{Theatre.read.show('08:30')}.to raise_error(RuntimeError, "you don't have enough money")}
+    it {expect{theatre.show('08:30')}.to change{theatre.money}.from(10).to(10 - AncientMovie::PRICE) }
```
Вместо этого матчера можно писать `.to change{theatre.money}.by(-AncientMovie::PRICE)`

zverok commented on task_6/theatre.rb:
```diff
@@ -0,0 +1,43 @@
+           end
+          end
+    time <= 9 ? ("0%d:00" %time) : ("%d:00" %time)
+  end
```
``` ruby
'%02d' % 8
# => "08" 
'%02d' % 12
# => "12" 
```

;)

zverok commented on task_6/theatre.rb:
```diff
@@ -0,0 +1,43 @@
+             else
+               raise ArgumentError, "film is not in the schedule"
+           end
+          end
```
А вот хорошо бы как-то придумать, чтобы эта логика только в одном месте повторялась, м? (В смысле, какой фильм какому периоду соответствует.) Если, например, соорудить хеш «период → фильтр» и немного пошаманить.

dmitrydi replied:
Я сваял нечто, но мне кажется, вышло тяжеловато. См. последние 2 коммита, я только theatre и theatre_spec выложил

zverok commented on task_6/theatre_spec.rb:
```diff
@@ -0,0 +1,26 @@
+require_relative '../theatre'
+
+describe Theatre do
```
Опять два раза файл со спеками?..

dmitrydi committed:
theatre_spec update 2

dmitrydi committed:
Theatre class changed

dmitrydi committed:
specs - last comments fixed

dmitrydi committed:
classes - last comments fixed

zverok commented on task_6/theatre.rb:
```diff
@@ -6,38 +6,28 @@ class Theatre < Cinema
 NOON = 12..17
 EVENING = 18..23
 NIGHT = 0..2
+FILTERS_FOR_PERIODS = { MORNING => {period: 'ancient'}, NOON => {genre: ['Comedy', 'Adventure']}, [EVENING, NIGHT] => {genre: ['Drama', 'Horror']} }
```
Не стесняйся такие штуки в несколько строк записывать:

``` ruby
FILTERS_FOR_PERIODS = {
  MORNING => {period: 'ancient'},
  NOON => {genre: ['Comedy', 'Adventure']},
  [EVENING, NIGHT] => {genre: ['Drama', 'Horror']}
}
```

читабельней же ж.

zverok commented on task_6/theatre.rb:
```diff
@@ -6,38 +6,28 @@ class Theatre < Cinema
-                   end
-    Theatre.new(list_to_show).pay(@money).filter_by_price(@money).collection.max_by{ |a| rand * a.rating }
+    hour = Time.parse(time).hour
+    filter = FILTERS_FOR_PERIODS.find{ |k, v| k.is_a?(Array) ? k.inject(false) { |m,a| m || a.include?(hour) } : k.include?(hour) }
```
мне кажется, что лучше было сделать фильтры более регулярными и тогда использовать их было б проще:

``` ruby
FILTERS_FOR_PERIODS = {
  MORNING => {period: 'ancient'},
  NOON => {genre: ['Comedy', 'Adventure']},
  EVENING => {genre: ['Drama', 'Horror']}, 
  NIGHT => {genre: ['Drama', 'Horror']}
}
```

zverok commented on task_6/theatre.rb:
```diff
@@ -6,38 +6,28 @@ class Theatre < Cinema
-    Theatre.new(list_to_show).pay(@money).filter_by_price(@money).collection.max_by{ |a| rand * a.rating }
+    hour = Time.parse(time).hour
+    filter = FILTERS_FOR_PERIODS.find{ |k, v| k.is_a?(Array) ? k.inject(false) { |m,a| m || a.include?(hour) } : k.include?(hour) }
+    raise ArgumentError, "The cinema is closed from 03:00 to 07:59" unless filter
```
Угу, но жёстко прописанное "03:00 to 07:59" нехорошо (а если требования поменяются? опять придётся в двух местах менять). 
Лучше просто текущее время написать.

zverok commented on task_6/theatre.rb:
```diff
@@ -6,38 +6,28 @@ class Theatre < Cinema
-           end
-          end
-    time <= 9 ? ("0%d:00" %time) : ("%d:00" %time)
+    period = FILTERS_FOR_PERIODS.find { |k,v| movie.any_match?(v.keys.first, v.values.first) }
```
да, вполне нормально. только метод я б назвал просто `match?`

dmitrydi replied:
`match?` уже есть просто. Я его немного доопределил и получил `any_match?`:

``` ruby
  def any_match?(key, val)
    return match?(key, val) unless val.is_a?(Array)
    key_val = self.send(key)
    val.inject(false) { |memo, v| memo || key_val.include?(v) }
  end
```

dmitrydi replied:
в итоге так:

``` ruby
  def match?(key, val)
    key_val = self.send(key)
    return val.inject(false) { |memo, v| memo || key_val.include?(v) } if val.is_a?(Array)
    key_val.is_a?(Numeric) ? (key_val == val) : key_val.include?(val)
  end
```

zverok replied:
Кстати, inject(false) тут очень большой перебор. Достаточно `val.any?{|v| key_val.include?(v) }`

zverok commented on task_6/netfix.rb:
```diff
@@ -2,8 +2,39 @@
-    self.filter(filter).filter_by_price(@money).collection.max_by{ |a| rand * a.rating }
+  def initialize(movie_array = nil)
+    super
+  	@money = 0
```
Чёт отступ превратился в тыкву.

zverok commented on task_6/netfix.rb:
```diff
@@ -2,8 +2,39 @@
+  def pay(amount)
+    if amount < 0
+      raise ArgumentError, "argument should be >=0"
+    else
```
Бубубу. else после raise не нужен, бубубу.

zverok commented on task_6/netfix.rb:
```diff
@@ -2,8 +2,39 @@
+  end
+
+  def select_movie(filters = nil)
+    filtered_collection = self.filter(filters).collection
```
Лишний self

dmitrydi committed:
trash

dmitrydi committed:
theatre modified

dmitrydi committed:
some minor changes about Theatre class
